routing/pipes.go

```diff
@@ -0,0 +1,97 @@
+package routing
+
```

Comment:
More comments needed in this file. Also, is this separate package currently a good idea?